### ISSUE TITLE
Fails to detect mounted partitions on target device

### ISSUE THREAD

**Issue by rugk** [coreos/coreos-installer#527]
# Bug #

I try to use this [to install it directly to an attached block devices as per the doc](https://docs.fedoraproject.org/en-US/fedora-coreos/bare-metal/#_installing_from_the_container), which will later be the boot medium.

However, it only succeeded for a few (2 IIRC) times. But now I tried to overwrite the device again – itering to find/try to fix the issue in my Ignition/CoreOs config. So I switched to downloading the image once, so it does not need to be redownloaded every time.
However, now it always fails with a non-explanatory error.

## Host Operating System Version ##

Fedora 33 Workstation

## Target Operating System Version ##

CoreOS v33, see below

## coreos-installer Version ##

```
$ coreos-installer --version
coreos-installer 0.9.0
$ dnf info coreos-installer
Last metadata expiration check: 0:00:19 ago on Mon 26 Apr 2021 11:28:46 PM CEST.
Installed Packages
Name         : coreos-installer
Version      : 0.9.0
Release      : 2.fc33
Architecture : x86_64
Size         : 6.3 M
Source       : rust-coreos-installer-0.9.0-2.fc33.src.rpm
Repository   : @System
From repo    : updates
Summary      : Installer for Fedora CoreOS and RHEL CoreOS
URL          : https://crates.io/crates/coreos-installer
License      : ASL 2.0 and MIT and zlib
Description  : coreos-installer installs Fedora CoreOS or RHEL CoreOS to bare-metal
             : machines (or, occasionally, to virtual machines).
```

## Expected Behavior ##

It should provide me a meaningful error… what is bad about a file existing?

## Actual Behavior ##

```shell
$ sudo coreos-installer install /dev/sdd -i ignition/[…].ign --image-file downloads/fedora-coreos-33.20210328.3.0-metal.x86_64.raw.xz
Copying image from downloads/fedora-coreos-33.20210328.3.0-metal.x86_64.raw.xz
Reading signature from downloads/fedora-coreos-33.20210328.3.0-metal.x86_64.raw.xz.sig
gpg: Signatur vom Mo 12 Apr 2021 19:26:31 CEST
gpg:                mittels RSA-Schlüssel 963A2BEB02009608FE67EA4249FD77499570FF31
gpg: Korrekte Signatur von "Fedora (33) <fedora-33-primary@fedoraproject.org>" [ultimativ]
> Read disk 542.7 MiB/542.7 MiB (100%)   
Writing Ignition config

Error: writing Ignition configuration

Caused by:
    0: opening destination Ignition config /tmp/coreos-installer-vpDQsB/ignition/config.ign
    1: File exists (os error 17)

Resetting partition table
Error: install failed
```

## Reproduction Steps ##
```sh
$ coreos-installer download -s stable -p metal -f raw.xz -C ./downloads               
zsh: correct 'download' to 'downloads' [nyae]? n
gpg: Signatur vom Mo 12 Apr 2021 19:26:31 CEST
gpg:                mittels RSA-Schlüssel 963A2BEB02009608FE67EA4249FD77499570FF31
gpg: Korrekte Signatur von "Fedora (33) <fedora-33-primary@fedoraproject.org>" [ultimativ]
> Read disk 542.7 MiB/542.7 MiB (100%)   
./downloads/fedora-coreos-33.20210328.3.0-metal.x86_64.raw.xz
$ sudo coreos-installer install /dev/sdd -i ignition/[…].ign --image-file downloads/fedora-coreos-33.20210328.3.0-metal.x86_64.raw.xz
# …
# as you can see above
```

## Other Information ##

As it takes so long, it seems only the Ignition embedding process fails.
(BTW, cant one manually trigger the embed without rewriting the whole downloaded image in this case?)

**Edit:** After formatting the device and re-trying it worked though. :slightly_smiling_face:

**Comment by bgilbert** [comment]
> Error: writing Ignition configuration
>
> Caused by:
>    0: opening destination Ignition config /tmp/coreos-installer-vpDQsB/ignition/config.ign
>    1: File exists (os error 17)

The error message is obscure because this situation should be impossible.  We've just written a fresh image, and when we go to write an Ignition config into it, somehow there's already one there.  I can think of a couple possibilities:

- Your `fedora-coreos-33.20210328.3.0-metal.x86_64.raw.xz` isn't actually a pristine image, and has a `config.ign` already embedded in it.  That seems unlikely, since your log shows a correct signature verification.
- We're trying to copy the Ignition config to a boot filesystem other than the one we just wrote.

Your reproduction steps didn't reproduce the problem for me.  Is there anything about your setup that makes either of the above cases more likely?  (For example, do you have another disk on the machine with CoreOS installed onto it?)

One thing you could try is dropping the `-i` option, letting install run to completion, and then manually mounting the `boot` partition from the target disk and seeing if it has an `ignition/config.ign` file.

> (BTW, cant one manually trigger the embed without rewriting the whole downloaded image in this case?)

Nope.  coreos-installer is meant to run to completion in one pass.

**Comment by rugk** [comment]
> Your fedora-coreos-33.20210328.3.0-metal.x86_64.raw.xz isn't actually a pristine image, and has a config.ign already embedded in it.

Yes, that’s very certainly not the case.

> We're trying to copy the Ignition config to a boot filesystem other than the one we just wrote.

Well… I used nearly the same commands for installing it for the first two times. So yes, the target device already has a CoreOS system on it. And yes, it also was booted once.
But I went back and just tried to overwrite it this way – which worked two times.

But as explained, now it failed (also tried two times) with the mentioned error.

> (For example, do you have another disk on the machine with CoreOS installed onto it?)

No.

> One thing you could try is dropping the -i option, letting install run to completion, and then manually mounting the boot partition from the target disk and seeing if it has an ignition/config.ign file.

Well… it would not embed my Ignition config then, obviously, which I want. :sweat_smile: 
Note that I just could solve it by formatting the target disk and then re-trying the command.
It worked flawlessly then.

**Comment by bgilbert** [comment]
> Well… I used nearly the same commands for installing it for the first two times. So yes, the target device already has a CoreOS system on it. And yes, it also was booted once.

Right, but we should have just overwritten that filesystem.

>> One thing you could try is dropping the `-i` option, letting install run to completion, and then manually mounting the boot partition from the target disk and seeing if it has an ignition/config.ign file.
>
> Well… it would not embed my Ignition config then, obviously, which I want.

Sure, but it would help troubleshoot the problem.

**Comment by rugk** [comment]
Okay, just encountered the issue again:

```shell
sudo coreos-installer install /dev/sdd -i ignition/[…].ign --image-file downloads/fedora-coreos-33.20210328.3.0-metal.x86_64.raw.xz --delete-karg 'console=ttyS0,115200n8'
[sudo] Passwort für rugk: 
Copying image from downloads/fedora-coreos-33.20210328.3.0-metal.x86_64.raw.xz
Reading signature from downloads/fedora-coreos-33.20210328.3.0-metal.x86_64.raw.xz.sig
gpg: Signatur vom Mo 12 Apr 2021 19:26:31 CEST
gpg:                mittels RSA-Schlüssel 963A2BEB02009608FE67EA4249FD77499570FF31
gpg: Korrekte Signatur von "Fedora (33) <fedora-33-primary@fedoraproject.org>" [ultimativ]
> Read disk 542.7 MiB/542.7 MiB (100%)   
Writing Ignition config

Error: writing Ignition configuration

Caused by:
    0: opening destination Ignition config /tmp/coreos-installer-6qJKy7/ignition/config.ign
    1: File exists (os error 17)

Resetting partition table
Error: install failed

```

So now I did what you said and dropped my ignition file:
```
sudo coreos-installer install /dev/sdd --image-file downloads/fedora-coreos-33.20210328.3.0-metal.x86_64.raw.xz --delete-karg 'console=ttyS0,115200n8' 
[sudo] Passwort für rugk: 
Copying image from downloads/fedora-coreos-33.20210328.3.0-metal.x86_64.raw.xz
Reading signature from downloads/fedora-coreos-33.20210328.3.0-metal.x86_64.raw.xz.sig
gpg: Signatur vom Mo 12 Apr 2021 19:26:31 CEST
gpg:                mittels RSA-Schlüssel 963A2BEB02009608FE67EA4249FD77499570FF31
gpg: Korrekte Signatur von "Fedora (33) <fedora-33-primary@fedoraproject.org>" [ultimativ]
> Read disk 542.7 MiB/542.7 MiB (100%)   
Modifying kernel arguments
Install complete.
```

Surprisingly, it succeeds. Maybe it does not want to overwrite the ignition file?

**Comment by rugk** [comment]
Ah I just found the cause!

It’s quite obvious, actually.

GNOME auto-mounts some disks when you enter a new drive. You have to unmount these, for it to work.
So mount the `/boot` and `/root` partitions before starting the installation and you should be able to reproduce this issue.

**Comment by rugk** [comment]
> >  (BTW, cant one manually trigger the embed without rewriting the whole downloaded image in this case?)
>
> Nope. coreos-installer is meant to run to completion in one pass.

Ah found it in `/boot/ignition/config.ign` - I could just be nasty and override that file… :stuck_out_tongue_winking_eye:

**Comment by bgilbert** [comment]
> GNOME auto-mounts some disks when you enter a new drive. You have to unmount these, for it to work.
> So mount the /boot and /root partitions before starting the installation and you should be able to reproduce this issue.

You are correct!  Thanks for finding this.  coreos-installer is supposed to fail if the disk has mounted partitions, but apparently that code broke at some point.

> Ah found it in `/boot/ignition/config.ign` - I could just be nasty and override that file…

Yeah, don't.  Your boot partition is no longer a pristine copy of the install image.  For now, just make sure the previous versions of the filesystems aren't mounted before installing, and we'll get coreos-installer fixed so that it fails in that case.

**Comment by bgilbert** [comment]
`BLKRRPART` is returning `0` rather than `-EBUSY` when there are partitions in use, so `Disk.get_busy_partitions()` succeeds when it should fail.  The bug was introduced in torvalds/linux@4601b4b130de2329fe06df80ed5d77265f2058e5 (backported to 5.10.20 and 5.11.3) and fixed in torvalds/linux@68e6582e8f2dc32fd2458b9926564faa1fb4560e (backported to 5.10.33 and 5.11.17).

FCOS `stable` is currently on 5.10.19, so shouldn't be affected, but FCOS `testing` and `next` are affected (5.11.15).  Fixed kernels are in Bodhi ([F33](https://bodhi.fedoraproject.org/updates/FEDORA-2021-55842d68c9), [F34](https://bodhi.fedoraproject.org/updates/FEDORA-2021-7a5e973e38)) so the next FCOS release cycle should have a fixed `next` and `testing` and a broken `stable`.  Since the fix is already propagating, I don't think it's worth adding a workaround to coreos-installer itself, and as to FCOS releases, I don't think the bug is significant enough to fast-track a kernel into the next `stable`.

I'll leave this open for now to track the fix.  Thanks for reporting!

**Comment by bgilbert** [comment]
This should be fixed in FCOS `next` 34.20210503.1.0.

**Comment by bgilbert** [comment]
This should be fixed in FCOS `testing` 34.20210518.2.0.

**Comment by bgilbert** [comment]
This should be fixed in FCOS `stable` 34.20210518.3.0.
